plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java

```diff
@@ -79,53 +73,66 @@ public static void addMavenLibraries(IProject project, List<Library> libraries,
+   * Returns the entries added to the classpath.
    */
-  static IClasspathEntry[] addLibraries(
-      IJavaProject javaProject, List<Library> libraries, IProgressMonitor monitor)
-          throws JavaModelException, CoreException {
-
-    // todo or we could grab the transitive dependency graph here if needed
-    // at this point we have the Java project so we could even filter
-    // entries already present in other libraries
+  public static IClasspathEntry[] addNativeLibrary(IJavaProject javaProject,
+      List<Library> libraries, IProgressMonitor monitor) throws CoreException {
     
-    return prepareLibraries(javaProject, libraries, monitor, true);
-  }
-
-  private static IClasspathEntry[] prepareLibraries(IJavaProject javaProject,
-      List<Library> libraries, IProgressMonitor monitor, boolean addToClasspath)
-          throws JavaModelException, CoreException {
     SubMonitor subMonitor = SubMonitor.convert(monitor,
         Messages.getString("adding.app.engine.libraries"), //$NON-NLS-1$
-        libraries.size() + 1); // + 1 because we pass the submonitor along below
-
-    List<IClasspathEntry> rawClasspath = Lists.newArrayList(javaProject.getRawClasspath());
+        2);
+    
+    Library masterLibrary = collectLibraryFiles(javaProject, libraries);
+    
+    List<IClasspathEntry> newEntries = computeEntries(javaProject, masterLibrary);
+    subMonitor.worked(1);
+    
+    ClasspathUtil.addClasspathEntries(javaProject.getProject(), newEntries, subMonitor);
+    runContainerResolverJob(javaProject);
+    
+    return newEntries.toArray(new IClasspathEntry[0]);
+  }
 
-    List<IClasspathEntry> newEntries = new ArrayList<>();
+  /**
+   * Adds all jars and dependencies from <code>libraries</code> to the master library.
+   * 
+   * @return the master library
+   */
+  public static Library collectLibraryFiles(IJavaProject javaProject, List<Library> libraries)
+      throws CoreException {
+    SortedSet<LibraryFile> masterFiles = new TreeSet<>();
     for (Library library : libraries) {
-      IClasspathEntry libraryContainer = makeClasspathEntry(library);
-      if (!rawClasspath.contains(libraryContainer)) {
-        newEntries.add(libraryContainer);
+      if (!library.isResolved()) {
+        library.resolveDependencies();
       }
-      subMonitor.worked(1);
+      masterFiles.addAll(library.getLibraryFiles());
     }
 
-    if (addToClasspath) {
-      ClasspathUtil.addClasspathEntries(javaProject.getProject(), newEntries, subMonitor);
-      runContainerResolverJob(javaProject);
-    }
-
-    return newEntries.toArray(new IClasspathEntry[0]);
+    Library masterLibrary = CloudLibraries.getMasterLibrary(javaProject);
+    masterFiles.addAll(masterLibrary.getLibraryFiles());
+    masterLibrary.setLibraryFiles(new ArrayList<LibraryFile>(masterFiles));
+    return masterLibrary;
   }
 
+  private static List<IClasspathEntry> computeEntries(IJavaProject javaProject, Library library)
```

Comment:
If we make this class only for the master library, this method becomes not so useful, because all it does is to return either 1) an empty list if `javaProject` already has the master container; or 2) a singleton list containing the master container. It would be clearer to rename this and add a Javadoc. As I said, all this refactoring can be done in another PR.

Reply:
again clearer once there's only a master container left